File: backend/app/services/spotify.py
import base64
import httpx

from backend.app.config import (
    SPOTIFY_CLIENT_ID,
    SPOTIFY_CLIENT_SECRET,
    SPOTIFY_REDIRECT_URI,
)

SPOTIFY_TOKEN_URL = "https://accounts.spotify.com/api/token"
SPOTIFY_API_URL = "https://api.spotify.com/v1"

async def get_access_token(code: str) -> dict:
    credentials = f"{SPOTIFY_CLIENT_ID}:{SPOTIFY_CLIENT_SECRET}"

    encoded_credentials = base64.b64encode(credentials.encode()).decode()

    headers = {
        "Authorization": f"Basic {encoded_credentials}",
        "Content-Type": "application/x-www-form-urlencoded",
    }

    data = {
        "grant_type": "authorization_code",
        "code": code,
        "redirect_uri": SPOTIFY_REDIRECT_URI,
    }

    async with httpx.AsyncClient() as client:
        response = await client.post(
            SPOTIFY_TOKEN_URL,
            headers=headers,
            data=data,
        )
        response.raise_for_status()
        return response.json()

File: frontend/src/services/spotify.ts
const API_URL = 'http://127.0.0.1:8000'

export async function getSpotifyAnalysis() {
  const response = await fetch(
    `${API_URL}/spotify/analysis`,
    {
      credentials: 'include',
    }
  )

  if (!response.ok) {
    throw new Error('Failed to fetch Spotify analysis')
  }

  return response.json()
}